### PR TITLE
Link jira tasks to a goal as a criteria?

### LINKED ISSUE #2484: Link jira tasks to a goal as a criteria?
1. When creating criteria, detect that it's a jira task
2. On backend, parse that task and update with info from jira
3. [#2843](https://github.com/taskany-inc/issues/issues/2843)

### LINKED ISSUE #2955: Check finished statuses of external task by add or update criteria
Make env vars those contains finished statuses from external service
Check external task for complete criteria by finished statuses

### PR DESCRIPTION
**Pull request by LamaEats**
## PR includes

- [x] Bug Fix
- [x] Feature

## Related issues

Resolve #2484 
Resolve #2955 

## QA Instructions, Screenshots, Recordings

image

### CODE REVIEW

File: src/utils/integration/jira.ts [hunk 1]
```diff
@@ -0,0 +1,226 @@
...
+            `Env variable \`JIRA_POSITIVE_STATUS_NAMES\` must be string, but get ${typeof process.env
+                .JIRA_POSITIVE_STATUS_NAMES}`,
+        );
+        assert(
+            process.env.JIRA_FINISHED_CATEGORY,
+            `Env variable \`JIRA_FINISHED_CATEGORY\` must be string, but get ${typeof process.env
+                .JIRA_FINISHED_CATEGORY}`,
+        );
+
+        this.positiveFinishedStatuses = process.env.JIRA_POSITIVE_STATUS_NAMES;
```
Comment (asabotovich) [inline src/utils/integration/jira.ts:104]: why it is not array?

File: src/utils/integration/jira.ts [hunk 2]
```diff
@@ -0,0 +1,226 @@
...
+        maxResults: params.limit,
+    });
+
+    return searchResults.issues.map((val: { fields: any }) => ({
+        ...val,
+        ...val.fields,
+    })) as Array<JiraIssue>;
+};
+
+jiraService.listStatus;
```
Comment (asabotovich) [inline src/utils/integration/jira.ts:226]: what does it mean?

File: src/utils/integration/jira.ts [hunk 3]
```diff
@@ -0,0 +1,226 @@
...
+        return this.makeUri(options);
+    }
+
+    protected getRequestHeaders(url: string, options?: JiraApi.UriOptions) {
+        // @ts-ignore
+        return this.makeRequestHeader(url, options);
+    }
+    /** end overriding private instamce methods */
+
+    public async statusesByCategory(categoryId: number) {
```
Comment (asabotovich) [inline src/utils/integration/jira.ts:139]: can't find usage

File: src/components/TaskBadge/TaskBadge.tsx [hunk 4]
```diff
@@ -0,0 +1,109 @@
...
+                nullable(state, (s) => <TaskBadgeState state={s.title} color={s.color} />),
+            )}
+            iconRight={children}
+            text={nullable(
+                href,
+                (h) => (
+                    <NextLink href={h} target="_blank" view="secondary" onClick={onClick}>
+                        {title}
+                        {nullable(project, (p) => (
+                            <>
```
Comment (asabotovich) [inline src/components/TaskBadge/TaskBadge.tsx:86]: unnecessary Fragment

File: src/components/CriteriaForm/CriteriaForm.tsx [hunk 5]
```diff
@@ -14,27 +15,39 @@ import {
...
+    type?: TaskTypeProps | null;
     _shortId: string;
 }
 
+const isGoalStateProps = (props: unknown): props is GoalStateProps =>
+    typeof props === 'object' && props != null && 'lightForeground' in props && 'darkForeground' in props;
+
+const isTaskStateProps = (props: unknown): props is TaskStateProps =>
+    typeof props === 'object' && props != null && 'src' in props && 'title' in props;
+
```
Comment (asabotovich) [inline src/components/CriteriaForm/CriteriaForm.tsx:44]: May it be better to have some `type` props to separate different types of SuggestItem?

File: src/utils/integration/jira.ts (hunk 1 above)
Comment (LamaEats) [reply src/utils/integration/jira.ts:104]: 'cause it is string `"\['value1', ...\]"` anyway

File: src/components/GoalCriteria/GoalCriteria.tsx [hunk 6]
```diff
@@ -56,25 +61,49 @@ interface GoalCriteriaProps extends CriteriaProps {
...
+        ownerEmail: string;
+        ownerName: string;
+    };
+}
+
+type UnionCriteria = CriteriaProps | GoalCriteriaProps | ExternalTaskCriteriaProps;
 
-interface GoalCriteriaEditableApi<T = UnionCriteria> {
+type CanBeNullableValue<T extends { [key: string]: any }> = {
+    [K in keyof T]: T[K] | null;
```
Comment (asabotovich) [inline src/components/GoalCriteria/GoalCriteria.tsx:90]: warning

File: src/components/GoalCriteriaSuggest.tsx [hunk 7]
```diff
@@ -68,21 +81,54 @@ export const GoalCriteriaSuggest: React.FC<GoalCriteriaSuggestProps> = ({
...
+            return goals.map(({ id, title, state, _shortId }) => ({
+                id,
+                title,
+                state:
+                    state?.lightForeground != null && state?.darkForeground != null
+                        ? {
+                              lightForeground: state.lightForeground,
+                              darkForeground: state.darkForeground,
+                          }
+                        : null,
```
Comment (asabotovich) [inline src/components/GoalCriteriaSuggest.tsx:113]: getStateProps?

File: src/components/GoalCriteria/GoalCriteria.tsx [hunk 8]
```diff
@@ -161,12 +273,31 @@ const GoalCriteria = ({ title, goal, weight, isDone }: Omit<GoalCriteriaProps, '
...
+
+        if (lightForeground && darkForeground) {
+            return {
+                lightForeground,
+                darkForeground,
+            };
+        }
+
+        return null;
+    }, [goal]);
```
Comment (asabotovich) [inline src/components/GoalCriteria/GoalCriteria.tsx:293]: getStateProps?

**Review by asabotovich — APPROVED** [review]

**Pull request merged**